### PR TITLE
[iOS] - Fix VPN subscriptions ignoring empty transactions states

### PR DESCRIPTION
**Pull request by Brandon-T**
## Summary
- When there's no transactions, we should fail instantly instead of doing nothing.

Resolves https://github.com/brave/brave-browser/issues/41012

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: ios/brave-ios/Sources/BraveVPN/Components/Subscription/BraveVPNInAppPurchaseObserver.swift [hunk 1]
```diff
@@ -115,7 +124,63 @@ public class BraveVPNInAppPurchaseObserver: NSObject, SKPaymentTransactionObserv
...
+            }
+          }
+
+          callPurchaseDelegateOnce = false
+        case .purchasing, .deferred:
+          Logger.module.debug("Received transaction state: purchasing")
+        case .failed:
+          Logger.module.debug("Received transaction state: failed")
+          self.delegate?.purchaseFailed(
+            error: .transactionError(error: transaction.error as? SKError)
```
Comment (kylehickinson) [inline ios/brave-ios/Sources/BraveVPN/Components/Subscription/BraveVPNInAppPurchaseObserver.swift:178]: do you not need to check `callPurchaseDelegateOnce` for the failure case?

Comment (Brandon-T) [reply ios/brave-ios/Sources/BraveVPN/Components/Subscription/BraveVPNInAppPurchaseObserver.swift]: Added.

File: ios/brave-ios/Sources/BraveVPN/Components/Subscription/BraveVPNInAppPurchaseObserver.swift [hunk 2]
```diff
@@ -115,7 +127,67 @@ public class BraveVPNInAppPurchaseObserver: NSObject, SKPaymentTransactionObserv
...
+            }
+          }
+
+          callPurchaseDelegateOnce = false
+        case .purchasing, .deferred:
+          Logger.module.debug("Received transaction state: purchasing")
+        case .failed:
+          Logger.module.debug("Received transaction state: failed")
+
+          if !callPurchaseDelegateOnce {
```
Comment (kylehickinson) [inline ios/brave-ios/Sources/BraveVPN/Components/Subscription/BraveVPNInAppPurchaseObserver.swift:181]: This is backwards, should be `if callPurchaseDelegateOnce`

**Review by kylehickinson — APPROVED** [review]

**Comment by kjozwiak** [comment]
Verification **PASSED** on `iPad Air (3rd Gen)` running `iOS 18.1` using the following build(s):

```
Brave | 1.73.52 Chromium: 131.0.6778.14 (Official Build) nightly (64-bit) 
--- | ---
Revision | 6996738c6195caf95549bd8a9bb48be180b4b62f
OS | iOS
```

Using the STR/Cases outlined via https://github.com/brave/brave-browser/issues/41012#issue-2520687063, ensured that clicking on `Restore` with a brand new account didn't let the user create a VPN profile even though a VPN premium purchase has never been made as per the following:

https://github.com/user-attachments/assets/763ebcc9-46db-4b4d-ad34-42d029fa4dcf

**Pull request merged**